filesystem/filesystem_test.go

```diff
@@ -109,6 +110,67 @@ func TestRemoveAllMetadata(t *testing.T) {
 	}
 }
 
+// Test that when MOUNTPOINT/.fscrypt is a pre-created symlink, fscrypt will
+// create/delete the metadata at the location pointed to by the symlink.
+//
+// This is a helper function that is called twice: once to test an absolute
+// symlink and once to test a relative symlink.
+func testSetupWithSymlink(t *testing.T, mnt *Mount, symlinkTarget string, realDir string) {
+	if err := os.Symlink(symlinkTarget, mnt.BaseDir()); err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(mnt.BaseDir())
+
+	if err := mnt.Setup(); err != nil {
+		t.Fatal(err)
+	}
+	defer mnt.RemoveAllMetadata()
+	if err := mnt.CheckSetup(); err != nil {
+		t.Fatal(err)
+	}
+	if !isSymlink(mnt.BaseDir()) {
+		t.Fatal("base dir should still be a symlink")
+	}
+	if !isDir(realDir) {
+		t.Fatal("real base dir should exist")
+	}
+	if err := mnt.RemoveAllMetadata(); err != nil {
+		t.Fatal(err)
+	}
+	if !isSymlink(mnt.BaseDir()) {
+		t.Fatal("base dir should still be a symlink")
+	}
+	if isDir(realDir) {
+		t.Fatal("real base dir should no longer exist")
+	}
+}
+
+func TestSetupWithAbsoluteSymlink(t *testing.T) {
+	mnt, err := getTestMount(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	realDir, err := ioutil.TempDir("", ".fscrypt-real")
```

Comment:
What's the reason for creating this directory just to remove it?

Reply:
Like in the test from https://github.com/google/fscrypt/pull/152, it's just supposed to be finding a temporary path that doesn't exist yet.  I've changed it to using a path inside the TempDir() instead.